primitives/trie/src/cache.rs

```diff
@@ -0,0 +1,864 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Trie Cache
+//!
+//! Provides an implementation of the [`TrieCache`](trie_db::TrieCache) trait.
+//! The implementation is split into three types [`SharedTrieCache`], [`LocalTrieCache`] and
+//! [`TrieCache`]. The [`SharedTrieCache`] is the instance that should be kept around for the entire
+//! lifetime of the node. It will store all cached trie nodes and values on a global level. Then
+//! there is the [`LocalTrieCache`] that should be kept around per state instance requested from the
+//! backend. As there are very likely multiple accesses to the state per instance, this
+//! [`LocalTrieCache`] is used to cache the nodes and the values before they are merged back to the
+//! shared instance. Last but not least there is the [`TrieCache`] that is being used per access to
+//! the state. It will use the [`SharedTrieCache`] and the [`LocalTrieCache`] to fulfill cache
+//! requests. If both of them don't provide the requested data it will be inserted into the
+//! [`LocalTrieCache`] and then later into the [`SharedTrieCache`].
+//!
+//! The [`SharedTrieCache`] is bound to some maximum number of bytes. It is ensured that it never
+//! runs above this limit. However as long as data is cached inside a [`LocalTrieCache`] it isn't
+//! taken into account when limiting the [`SharedTrieCache`]. This means that for the lifetime of a
+//! [`LocalTrieCache`] the actual memory usage could be above the allowed maximum.
+
+use crate::{Error, NodeCodec};
+use hash_db::Hasher;
+use lru::LruCache;
+use nohash_hasher::{BuildNoHashHasher, IntMap, IntSet};
+use parking_lot::{Mutex, MutexGuard, RwLock, RwLockReadGuard};
+use std::{
+	collections::{hash_map::Entry, HashMap, HashSet},
+	hash::{BuildHasher, Hasher as _},
+	mem,
+	sync::Arc,
+};
+use trie_db::{node::NodeOwned, CachedValue};
+
+const LOG_TARGET: &str = "trie-cache";
+
+lazy_static::lazy_static! {
+	static ref RANDOM_STATE: ahash::RandomState = ahash::RandomState::default();
+}
+
+/// No hashing [`LruCache`].
+///
+/// The key is an `u64` that is expected to be unique.
+type NoHashingLruCache<T> = lru::LruCache<u64, T, BuildNoHashHasher<u64>>;
+
+/// Returns a hasher prepared to build the final value cache key.
+///
+/// To build the final hash the storage key still needs to be added to the hasher.
+/// If you are only interested in the full value cache key, you can directly use
+/// [`value_cache_get_key`].
+fn value_cache_partial_hash(storage_root: &impl AsRef<[u8]>) -> impl std::hash::Hasher + Clone {
+	let mut hasher = RANDOM_STATE.build_hasher();
+	hasher.write(storage_root.as_ref());
+	hasher
+}
+
+/// Get the key for the value cache.
+///
+/// The key is build by hashing the actual storage `key` and the `storage_root`.
+fn value_cache_get_key(key: &[u8], storage_root: &impl AsRef<[u8]>) -> u64 {
```

Comment:
What happens if there's a collision here?

Reply:
That is a good question.. :see_no_evil:

Reply:
Wasn't it already questioned in https://github.com/paritytech/substrate/pull/11407#discussion_r872137237 . I must admit not liking the conclusion so much

Reply:
After my first initial review I assumed this 64 bit value is used as a hash. I.e. an actual key would be compared if there's a hash match. But now I can see that this value is actually used as a key and no additional checks on the actual key are performed. Even though the chance is low it is still high enough to cause random consensus issues over thousand of nodes. This should be 256  bits with DDoS resistant hash function, as @koute and @cheme have already mentioned in the comment mentioned above.

Reply:
I'm already reworking this. To ensure we don't fuck it up :)

Reply:
@cheme @arkpar please take another look. Sadly the reading of the entire state is now slower. But maybe that is because of the cache size. I will need to check this tomorrow. However, the other benchmarks actually got faster.